enhancements/project-request-limit.md

```diff
@@ -0,0 +1,392 @@
+
+
+### Goals
+1. Enable the `requestlimit.project.openshift.io/ProjectRequestLimit` admission plugin.
+2. Use existing extension points, libraries, and installation mechanisms in the manner we would recommend to external teams.
+3. Have a fairly straightforward way to install and enable this admission plugin.
+
+### Non-Goals
+1. Revisit how `requestlimit.project.openshift.io/ProjectRequestLimit` works. We want to lift it as is.
+2. Couple a slow moving admission plugin to a fast moving `kube-apiserver`.
+3. Since `Project` type is owned by `openshift-apiserver`, rebootstrapping is not applicable.
+
+### Open Questions
+1. How do we describe version skew limitations to OLM so our operator gets uninstalled *before* an illegal downgrade or
+ upgrade? This is a concrete case of the API we want to use isn't available before 1.16 and after 1.18, the previous API could be gone.
+
+## Proposal
+1. Create a `Validating` admission webhook server that provides `requestlimit.project.openshift.io/ProjectRequestLimit`.
+2. The admission webhook will be fronted by a `Service`.
+3. The webhook server is reachable through the `kube-apiserver` via the `kubernetes.default.svc service` service. 
+2. Create an operator that manages all lifecycle aspects of the admission webhook via a custom resource.
+3. Productize the operator and ship it as a `RedHat operator` via `OperatorHub`. 
+4. Provide official documentation on how to interact with the `ProjectRequestLimit` operator.
+
+### User Stories [optional]
+*Story 1*: As a cluster admin I want to enable `ProjectRequestLimit` on an OpenShift 4.x cluster so that I can limit creation of `Project`.
+
+*Story 2*: As a cluster admin I want to remove `ProjectRequestLimit` from my cluster.
+
+*Story 3*: As a cluster admin I want to be able to install `ProjectRequestLimit` in a disconnected environment.
+
+*Story 4*: As a cluster admin I want my cluster to automatically upgrade to the new version of `ProjectRequestLimit` when available.
+
+*Story 5*: As a cluster admin I want to be able to specify (at any time) limits on the number of user provisioned `Project(s)`.
+
+Notes:
+* Port `ProjectRequestLimit` plugin from 3.11 into a `Validating` admission webhook.
+* The source code is here: https://github.com/openshift/origin/blob/release-3.11/pkg/project/apiserver/admission/requestlimit/admission.go
+* Use  [generic-admission-server](https://github.com/openshift/generic-admission-server) to wire the validating admission webhook server. 
+* The operator is OLM enabled.
+* Ship `ProjectRequestLimit` as a RedHat operator via OperatorHub.
+* For productization, if we go through ART, integration work should target 4.5.
+* The operator should wire the OLM manifests accordingly to facilitate disconnected install (`relatedImages`).
+
+
+### Implementation Details/Notes/Constraints
+
+#### The Operand
+We build the webhook admission server to also be an extension API server, it becomes possible to aggregate it as a normal
```

Comment:
"If ..." missing?

Reply:
clarified it. what I meant to say `thus it will enable us to aggregate it as a normal
API server`